subsys/bluetooth/services/bms.c

```diff
@@ -0,0 +1,434 @@
+	BT_GATT_BMS_OP_DEL_ALL_BONDS = 0x06,
+
+	/* Initiates the procedure to delete all bonds except for the one of
+	 * the requesting device
+	 */
+	BT_GATT_BMS_OP_DEL_ALL_BUT_ACTIVE_BOND = 0x09,
+};
+
+/* State of BMS instance */
+enum bt_gatt_bms_state {
+	/* Idle */
+	IDLE,
+
+	/* Control point operation in progress */
+	OPERATION_IN_PROGRESS,
+
+	/* Bond deletion operation in progress */
+	DELETION_IN_PROGRESS,
+};
+
+/* Execution descriptor */
+struct bt_gatt_bms_bond_add_exec {
+	/* Kernel work item */
+	struct k_work work;
+
+	/* Bond operation code */
+	enum bt_gatt_bms_op op_code;
+
+	/* Requestor address */
+	bt_addr_le_t req_addr;
+};
+
+static struct bt_gatt_bms {
+	enum bt_gatt_bms_state state;
+	struct bt_gatt_bms_bond_add_exec bond_add_exec;
+	struct k_work bond_del_exec;
+	struct bt_gatt_bms_cb cbs;
+	struct bt_gatt_bms_features features;
+	bt_addr_le_t bonded_addrs[CONFIG_BT_MAX_PAIRED];
+} bms_inst;
+
+static inline bool is_bond_item_free(const bt_addr_le_t *addr)
+{
+	const bt_addr_le_t free_addr = {0};
+
+	return (memcmp(&free_addr, addr, sizeof(free_addr)) == 0);
+}
+
+static void bond_item_add(const bt_addr_le_t *addr)
+{
+	s32_t free_item_index = -1;
+
+	for (u32_t i = 0; i < ARRAY_SIZE(bms_inst.bonded_addrs); i++) {
+		if (bt_addr_le_cmp(addr, &bms_inst.bonded_addrs[i]) == 0) {
+			LOG_DBG("This bonded address is already tracked");
+			return;
+		}
+
+		if (free_item_index == -1) {
+			if (is_bond_item_free(&bms_inst.bonded_addrs[i])) {
+				free_item_index = (s32_t) i;
+			}
+		}
+	}
+
+	if (free_item_index == -1) {
+		LOG_ERR("No space left to track a new bonded address");
+		return;
+	}
+
+	memcpy(&bms_inst.bonded_addrs[free_item_index], addr, sizeof(*addr));
+}
+
+static void bond_item_remove(bt_addr_le_t *addr)
+{
+	int err;
+	struct bt_conn *conn;
+	char addr_str[BT_ADDR_LE_STR_LEN];
+
+	if (is_bond_item_free(addr)) {
+		return;
+	}
+
+	bt_addr_le_to_str(addr, addr_str, sizeof(addr_str));
+
+	conn = bt_conn_lookup_addr_le(BT_ID_DEFAULT, addr);
+	if (conn) {
+		LOG_DBG("Bond information for: %s will be removed on "
+			"disconnect", addr_str);
+
+		bt_conn_unref(conn);
+	} else {
+		LOG_INF("Removing bond information for: %s", addr_str);
+
+		err = bt_unpair(BT_ID_DEFAULT, addr);
+		if (err) {
+			LOG_ERR("Unable to remove bond information: %d", err);
+		}
+
+		memset(addr, 0, sizeof(*addr));
+	}
+}
+
+static void bond_items_remove(void)
+{
+	for (u32_t i = 0; i < ARRAY_SIZE(bms_inst.bonded_addrs); i++) {
+		bond_item_remove(&bms_inst.bonded_addrs[i]);
+	}
+}
+
+static void on_del_all_bonds_iterate(const struct bt_bond_info *info,
+				     void *user_data)
+{
+	bond_item_add(&info->addr);
+}
+
+static void on_del_rest_bonds_iterate(const struct bt_bond_info *info,
+				      void *user_data)
+{
+	const bt_addr_le_t *req_addr = (const bt_addr_le_t *) user_data;
+
+	if (bt_addr_le_cmp(&info->addr, req_addr) != 0) {
+		bond_item_add(&info->addr);
+	}
+}
+
+static void bond_add_execute(struct k_work *item)
```

Comment:
This one is a bit hard to understand. The function is add execute, but it mixes deletions. It calls "on_del_all_bonds_iterate" (which calls bond_item_add, so that seems fine, but the name is weird).
And then it sets the state "DELETION_IN_PROGRESS".

Reply:
I changed naming for bond iterators and added block comments describing what happens in this function.